New SSL instructions for FIPS compatibility (#1806)
neo4j/docs-operations: modules/ROOT/pages/security/ssl-fips-compatibility.adoc

[role=enterprise-edition]
[[ssl-fips-compatibility]]
= Configuring SSL for FIPS 140-2 compatibility
:description: How to configure Neo4j to use FIPS compatible SSL encryption.
:keywords: ssl, tls, authentication, encryption, encrypted, security, fips, fips 140, fips 140-2, nist, hipaa

Federal Information Processing Standards (FIPS) 140 is a U.S. government standard established by the National Institute of Standards and Technology (NIST) which is used to accredit cryptographic modules such as those used in TLS network encryption.
While FIPS 140 compliance is primarily required for federal agencies and their contractors, it also is used in the healthcare sector under regulations like the Health Insurance Portability and Accountability Act (HIPAA) to protect patient data.

> This is a guide to help configure Neo4j to use TLS / SSL encryption in a FIPS-compatible way.

Reviewer: ```suggestion
This guide helps configure Neo4j to use TLS/SSL encryption in a FIPS-compliant way.
```

It is supplementary to the xref:security/ssl-framework.adoc[] documentation, as many of the configuration processes and requirements are the same.

[[ssl-fips-prerequisites]]
== Prerequisites

* Verify that the machine running Neo4j has FIPS-compatible hardware and operating system.
Only xref:installation/requirements.adoc#deployment-requirements-software[Linux operating systems] are supported for Neo4j FIPS compatibility at this time.
* Use Neo4j Enterprise 5.23.0 or later.
* Install and configure a non-native authentication provider, for example LDAP or SSO. See xref:authentication-authorization/index.adoc[].

[[fips-ssl-provider-docker]]
== Enable FIPS SSL provider (Docker)

The Neo4j RedHat UBI9 Docker image comes with the SSL provider and dependencies pre-installed, but it is not enabled by default.

[NOTE]
====
The Debian based Neo4j Docker image does *not* support FIPS compatible encryption.
====

To enable the OpenSSL FIPS provider, set the environment variable `NEO4J_OPENSSL_FIPS_ENABLE=true` when starting the container.

[source, console, subs="attributes"]

> .Example starting a Neo4j UBI9 container with FIPS enable flag set.

Reviewer: ```suggestion
.Example of starting a Neo4j UBI9 container with FIPS enable flag set.
```

----
docker run -it --rm \
    --publish=7474:7474 \
    --publish=7687:7687 \
    --env=NEO4J_OPENSSL_FIPS_ENABLE=true \
    --volume=$HOME/neo4j/data:/data \
    --volume=$HOME/neo4j/conf:/conf \
    --volume=$HOME/neo4j/certificates:/ssl \
neo4j:{neo4j-version-exact}-enterprise-ubi9
----

[[fips-ssl-provider]]
== Enable FIPS SSL provider

[IMPORTANT]
====
Skip this section if using Neo4j in Docker.
====

The secure networking in Neo4j is provided through the Netty library, which supports both the native JDK SSL provider and Netty-supported OpenSSL derivatives.
Specifically Netty's _Forked Tomcat Native_ library called https://github.com/netty/netty-tcnative[netty-tcnative].

The `netty-tcnative` library is provided in several variants.
However, to achieve FIPS compliance, you must use the dynamically linked version of `netty-tcnative` alongside a FIPS-compatible installation of OpenSSL.

The dynamically linked library requires the following dependencies to be installedfootnote:[https://netty.io/wiki/forked-tomcat-native.html]:

* Apache Portable Runtime Library
* A FIPS certified version of OpenSSL, with a FIPS provider installed and set as default.

> Refer to https://netty.io/wiki/forked-tomcat-native.html/[Forked Tomcat Native] for more information.

Reviewer: ```suggestion
Refer to https://netty.io/wiki/forked-tomcat-native.html[Forked Tomcat Native] for more information.
```

Select the one matching the local machine's operating system and architecture.
+
. Verify the dependencies are correctly installed using https://www.man7.org/linux/man-pages/man1/ldd.1.html[`ldd`]:
+
[source, console]
.Verify netty-tcnative dependencies are installed
----
unzip -d /tmp <NEO4J_LIB>/netty-tcnative/netty-tcnative-*-linux-$(arch).jar
ldd /tmp/META-INF/native/libnetty_tcnative_linux_*.so
rm -rf /tmp/META-INF
----
+
[source, console]
.Verify Fedora variant of netty-tcnative dependencies are installed
----
unzip -d /tmp <NEO4J_LIB>/netty-tcnative/netty-tcnative-*-linux-$(arch)-fedora.jar
ldd /tmp/META-INF/native/libnetty_tcnative_linux_$(arch).so
rm -rf /tmp/META-INF
----
The `ldd` command shows a list of library dependencies and where they will be loaded from on the local machine.
** If any dependencies are missing, they must be installed or Neo4j will fail to run.
** The `libssl.so` and `libcrypto.so` libraries listed must be the ones installed with OpenSSL in the previous steps.
+
. Copy the verified jar to _<NEO4J_LIB>_.
+
[NOTE]
====
Only copy *one* of the jars, otherwise Neo4j will not be able to resolve dependencies at runtime.

> The error, if this happens, contains a message like:

Reviewer: ```suggestion
In case of the error, you get a message like:
```
